### DISCUSSION TITLE
What is the principle of trivy?

### DISCUSSION THREAD

**Discussion by fighterhit** [aquasecurity/trivy#1471] [Q&A] [answered]
What is the principle of trivy? For example, why can it detect the latest log4j2 vulnerability in the container image? Will it download the image, decompress and scan all the jar packages in the image?

**Comment by AndreyLevchenko** [comment, marked as answer]
Hi @fighterhit 
Thank you for your interest and yes,  you are correct

**Reply by fighterhit** [reply]
> Hi @fighterhit Thank you for your interest and yes, you are correct

Hi @AndreyLevchenko , thanks for your reply. However, when the image is being scanned or after it is completed, I execute `docker images`, but the image is not found. Is the image deleted immediately after detection?

**Reply by AndreyLevchenko** [reply]
if image is found locally it will be scanned, otherwise Trivy works directly with remote docker registry.

**Reply by fighterhit** [reply]
> if image is found locally it will be scanned, otherwise Trivy works directly with remote docker registry.

Thanks @AndreyLevchenko ,I am curious about how the image is analyzed when it is in the remote registry. Is it downloaded layer by layer to the local and then analyzed after decompression?

**Reply by knqyf263** [reply]
>Is it downloaded layer by layer to the local and then analyzed after decompression?

Yes

**Reply by fighterhit** [reply]
> > Is it downloaded layer by layer to the local and then analyzed after decompression?
> 
> Yes

Thanks @knqyf263 , since it has been downloaded, why can't I see it when I use `docker images`? Was it deleted after the scan was completed?

**Reply by knqyf263** [reply]
Trivy downloads layers directly without Docker Engine. Each layer is processed in memory.

**Reply by fighterhit** [reply]
> Trivy downloads layers directly without Docker Engine. Each layer is processed in memory.

Got it @knqyf263 , thanks!
